frame/executive/src/lib.rs

```diff
@@ -355,6 +357,18 @@ where
 
 		// post-extrinsics book-keeping
 		<frame_system::Module<System>>::note_finished_extrinsics();
+
+		let weight =  <frame_system::Module<System>>::block_weight();
+		let max_weight =  <System::BlockWeights as frame_support::traits::Get<_>>::get().max_block;
+		let mut remaining_weight = max_weight.saturating_sub(weight.total());
+
+		if remaining_weight > 0 {
+			let mut reduced_weight = <frame_system::Module<System> as OnIdle<System::BlockNumber>>::on_idle(block_number, remaining_weight);
+			remaining_weight = remaining_weight.saturating_sub(reduced_weight);
+			reduced_weight = <AllModules as OnIdle<System::BlockNumber>>::on_idle(block_number, remaining_weight).saturating_add(reduced_weight);
+			<frame_system::Module::<System>>::register_extra_weight_unchecked(reduced_weight, DispatchClass::Mandatory);
+		}
```

Comment:
```suggestion
		if remaining_weight > 0 {
			let mut used_weight = <frame_system::Module<System> as OnIdle<System::BlockNumber>>::on_idle(block_number, remaining_weight);
			remaining_weight = remaining_weight.saturating_sub(used_weight);
			used_weight = <AllModules as OnIdle<System::BlockNumber>>::on_idle(block_number, remaining_weight).saturating_add(used_weight);
			<frame_system::Module::<System>>::register_extra_weight_unchecked(used_weight, DispatchClass::Mandatory);
		}
```

Reply:
Also I'm unsure about the class being `Mandatory`. @thiolliere ?

Reply:
I think Mandatory makes sense as it is a weight used by the chain and not by operational or normal extrinsics